### PR TITLE
added userPass authentication to the hashicorp vault provider

### LINKED ISSUE #2538: Add UserPass authentication method to hasicorp vault provider
**Is your feature request related to a problem? Please describe.**
The authentication method userPass is missing, sadly the SaaS provider which I use in my project only support this method.

**Describe the solution you'd like**
Adding the support for the userPass authentication method according to the already supported ones. 

**Describe alternatives you've considered**
In the future the provider will add different authentication methods, nervertheless it would be good to support userPass here too

**Additional context**
https://developer.hashicorp.com/vault/docs/auth/userpass

### PR DESCRIPTION
**Pull request by c0ffee**
## Problem Statement

Adding UserPass authentication to the vault provider

## Related Issue

Fixes #2538

## Proposed Changes

Adding support for this authentication method accordingly to the already existing authentication methods

## Checklist

- [x] I have read the [contribution guidelines](https://external-secrets.io/latest/contributing/process/#submitting-a-pull-request)
- [x] All commits are signed with `git commit --signoff`
- [x] My changes have reasonable test coverage
- [x] All tests pass with `make test`
- [x] I ensured my PR is ready for review with `make reviewable`

### CODE REVIEW

**Comment by c0ffee** [comment]
sadly i am not able to get "make reviewable" to run, but i get the same result on main, is it in general broken?

I0725 14:53:43.857769   86911 main.go:167] written to /home/ubuntu/external-secrets/hack/api-docs/../..//docs/api/spec.md
mkdir -p /home/ubuntu/external-secrets/hack/api-docs/../..//site
docker run \
    --mount type=bind,source=/home/ubuntu/external-secrets/hack/api-docs/../../,target=/repo \
	--sig-proxy=true \
	--rm \
	--user 1000:1000 \
	--env "GIT_COMMITTER_NAME=" \
	--env "GIT_COMMITTER_EMAIL=" \
	github.com/external-secrets-mkdocs:latest \
	/bin/bash -c "cd /repo && mike deploy --ignore --update-aliases -F hack/api-docs/mkdocs.yml main unstable;"
Config value 'nav': Expected nav to be a list, got dict with keys ('v2', 'v1')
INFO     -  DeprecationWarning: pkg_resources is deprecated as an API
  File "/usr/lib/python3.11/site-packages/mkdocs_macros/context.py", line 19, in <module>
    import pkg_resources
  File "/usr/lib/python3.11/site-packages/pkg_resources/__init__.py", line 121, in <module>
    warnings.warn("pkg_resources is deprecated as an API", DeprecationWarning)

WARNING  -  Config value 'nav': Expected nav to be a list, got dict with keys ('v2', 'v1')
INFO     -  [macros] - Macros arguments: {'module_name': 'main', 'modules': [], 'include_dir': 'docs/snippets', 'include_yaml': [], 'j2_block_start_string': '', 'j2_block_end_string': '', 'j2_variable_start_string': '', 'j2_variable_end_string': '', 'on_undefined': 'keep', 'on_error_fail': False, 'verbose': False}
INFO     -  [macros] - Extra variables (config file): ['version', 'analytics']
INFO     -  [macros] - Extra filters (module): ['pretty']
INFO     -  [macros] - Includes directory: docs/snippets
INFO     -  Cleaning site directory
INFO     -  Building documentation to directory: /repo/site
INFO     -  The following pages exist in the docs directory, but are not included in the "nav" configuration:
  - spec.md
  - provider/aws-pushsecret.md
  - snippets/provider-aws-access.md
INFO     -  [macros] - ERROR # _Macro Syntax Error_
_Line 243 in Markdown file:_ **unexpected '.'**
```python
        secret: "{{ .password }}"
```
INFO     -  Documentation built in 11.41 seconds
error: error getting config 'user.name'
make[1]: *** [Makefile:50: build] Error 1
make[1]: Leaving directory '/home/ubuntu/external-secrets/hack/api-docs'
make: *** [Makefile:196: docs] Error 2

**Comment by arnouthoebreckx** [comment]
Hey @c0ffee, I had the same issue it means that you've not properly setup your Git settings

git config --global user.name "..."
git config --global user.email "..."

Fixed the "make reviewable" for me. You also have to install helm etc.

**Comment by c0ffee** [comment]
Hey @arnouthoebreckx ,

thank you for the hints! now i could finally run it :)
maybe we should add this to the dev doku

**Comment by c0ffee** [comment]
can one of the maintainers pleave have a look :)

File: pkg/provider/vault/vault.go [hunk 1]
```diff
@@ -1193,6 +1208,18 @@ func setLdapAuthToken(ctx context.Context, v *client) (bool, error) {
 	return false, nil
 }
 
+func setUserPassAuthToken(ctx context.Context, v *client) (bool, error) {
+	userPassAuth := v.store.Auth.UserPass
+	if userPassAuth != nil {
+		err := v.requestTokenWithUserPassAuth(ctx, userPassAuth)
+		if err != nil {
+			return true, err
```
Comment (moolen) [inline pkg/provider/vault/vault.go:1216]: Why return true here, is there a need for it? From what i can see in `requestTokenWithUserPassAuth` it is either an error or nil if successful.

**Review by moolen — COMMENTED** [review]
Sorry, i haven't yet found the time to respond earlier. Thanks for pinging, that definitely brings it up front in my workqueue :+1:!

Changes LGTM. One nit, see comment. I haven't had the time yet to do a full end to end test, will do that asap.

**Review by moolen — APPROVED** [review]
Did some manual tests using `userpass` auth, looks good to me!
Just that open question, otherwise this is ready to be merged.
Thank you for your contribution :bow:

File: pkg/provider/vault/vault.go (hunk 1 above)
Comment (c0ffee) [reply pkg/provider/vault/vault.go:1216]: well good point, i only tried to to be constant with the other functions which are all doing it like this ;)
it is used to signal that a token exists:
        tokenExists, err = setUserPassAuthToken(ctx, v)

but yeah if i change it we have to change the whole function: https://github.com/external-secrets/external-secrets/blob/80d48c44e1300a4e86e57886d95abfb44f45875e/pkg/provider/vault/vault.go#L1101

Comment (moolen) [reply pkg/provider/vault/vault.go:1216]: I just meant that the return `true` is odd, when you receive an error. From what i can see in the `requestTokenWithUserPassAuth` there seems to be no case where you return an error AND a token was created.

Hence we can simply `return false, err` here.
But you're right, that way it's consistent with the other funcs, which all should be changed :sweat_smile: 

We'll leave that for a different PR :)

**Comment by moolen** [comment]
Thanks @c0ffee for your contribution, much appreciated! :medal_sports:

**Pull request merged**